src/main/java/com/aphyr/riemann/client/LocalhostResolver.java

```diff
@@ -0,0 +1,93 @@
+package com.aphyr.riemann.client;
+
+import java.net.UnknownHostException;
+
+/**
+ * A "Smarter" localhost resolver
+ * see issue: https://github.com/aphyr/riemann-java-client/issues/44
+ * Trying to avoid a lot of calls to java.net.InetAddress.getLocalHost()
+ * which under AWS trigger DNS resolving and have relatively high latency *per event*
+ * usually, the hostname doesn't change so often to warrant a real query.
+ *
+ * A real call to java.net.InetAddress.getLocalHost().getHostName()
+ * is made only if:
+ * 1) the refresh interval has passed (=result is stale)
+ * AND
+ * 2) no env vars that identify the hostname are found
+ */
+public class LocalhostResolver {
+
+    // default hostname env var names on Win/Nix
+    public static final String COMPUTERNAME = "COMPUTERNAME"; // Windows
+    public static final String HOSTNAME = "HOSTNAME"; // Nix
+
+    // how often should we refresh the cached hostname
+    public static long refreshIntervalMillis = 60 * 1000;
+
+    // cached hostname result
+    private static boolean envResolved = false;
+    private static String hostname;
+
+    // update (refresh) time management
+    private static long lastUpdate = 0;
+    public static long getLastUpdateTime() { return lastUpdate; }
+    // this is mostly for testing, plus ability to force a refresh
+    public static void setLastUpdateTime(long time) {
+        lastUpdate = time;
+    }
+
+    static {
+        resolveByEnv();
+    }
+
+    /**
+     * get resolved hostname.
+     * encapsulates all lookup and caching logic.
+     *
+     * @return the hostname
+     */
+    public static String getResolvedHostname() {
+        long now = System.currentTimeMillis();
+        if((now - refreshIntervalMillis) > lastUpdate) {
+            refreshResolve();
+        }
+
+        return hostname;
+    }
+
+    /**
+     * forces a new resolve even if refresh interval has not passed yet
+     */
+    public static void refreshResolve() {
```

Comment:
DNS resolution no doubt trumps synchronization, but if I'm not mistaken, test+set of `lastUpdate` against the interval also needs to be synchronized if we're to prevent parallel DNS lookups, and that locking will happen each call to `getResolvedHostname` (on each event basically). Code that runs each event should be as fast as possible.
I feel bad paying (even small sums) each event to save on the rare cases of collision, not to mention introducing added system complexity that comes with locking.

The concern is to prevent parallel resolutions, so one timer thread do it, no locking. simpler and cleaner, no?
That timer, internal to the resolver, firing and resolving on each interval, while callers always get the most recent cached hostname value.